resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeValidator.scala

```diff
@@ -0,0 +1,193 @@
+   *
+   * Please note that if multiple resources are defined like described above,
+   * the error messages will be concatenated.<br>
+   * Example of such a config:<br>
+   * - spark.yarn.driver.resource.memory=2g<br>
+   * - spark.yarn.executor.resource.cores=2<br>
+   * Then the following two error messages will be printed:<br>
+   * - "memory cannot be requested with config spark.yarn.driver.resource.memory,
+   * please use config spark.driver.memory instead!<br>
+   * - "cores cannot be requested with config spark.yarn.executor.resource.cores,
+   * please use config spark.executor.cores instead!<br>
+   *
+   * @param sparkConf
+   */
+  def validateResources(sparkConf: SparkConf): Unit = {
+    val requestedResources = new RequestedResources(sparkConf)
+    val sb = new mutable.StringBuilder()
+    POSSIBLE_RESOURCE_DEFINITIONS.foreach { rcp =>
+      val customResources: Map[String, String] = getCustomResourceValue(requestedResources, rcp)
+      val (standardResourceConfigKey: String, customResourceConfigKey: String) =
+        getResourceConfigKeys(rcp)
+
+      val errorMessage =
+        if (customResources.contains(customResourceConfigKey)) {
+          s"${rcp.resourceType} cannot be requested with config $customResourceConfigKey, " +
+              s"please use config $standardResourceConfigKey instead!"
+        } else {
+          ""
+        }
+      if (errorMessage.nonEmpty) {
+        printErrorMessageToBuffer(sb, errorMessage)
+      }
+    }
+
+    if (sb.nonEmpty) {
+      throw new SparkException(sb.toString())
+    }
+  }
+
+  /**
+   * Returns the requested map of custom resources,
+   * based on the ResourceConfigProperties argument.
+   * @return
+   */
+  private def getCustomResourceValue(
+      requestedResources: RequestedResources,
+      rcp: ResourceConfigProperties): Map[String, String] = {
+    var customResources: Map[String, String] = null
+    (rcp.processType, rcp.runMode, rcp.resourceType) match {
+      case (ProcessType.EXECUTOR, _, _) => customResources =
+          requestedResources.customExecutorResources
+      case (ProcessType.AM, Some(RunMode.CLIENT), _) => customResources =
+          requestedResources.customAMResources
+      case (ProcessType.DRIVER, Some(RunMode.CLUSTER), _) => customResources =
+          requestedResources.customDriverResources
+      case _ => throw new SparkException(s"ResourceConfigProperties contains an " +
+          s"invalid combination of fields: $rcp")
+    }
+    customResources
+  }
+
+  /**
+   * Returns a tuple of standard resource config key and custom resource config key, based on the
+   * processType and runMode fields of the ResourceConfigProperties argument.
+   * Standard resources are memory and cores.
+   *
+   * @param rcp
+   * @return
+   */
+  private def getResourceConfigKeys(rcp: ResourceConfigProperties): (String, String) = {
+    val standardResourceConfigKey: String = if (rcp.processType == ProcessType.AM) {
```

Comment:
You can omit types on local vars/vals ; only in cases where it's unclear or ambiguous is it really needed.

Reply:
Good to know, removed the type from here.